src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -59,28 +58,14 @@ private BoundExpression RewriteCollectionExpressionConversion(Conversion convers
                     case CollectionExpressionTypeKind.Span:
                     case CollectionExpressionTypeKind.ReadOnlySpan:
                         Debug.Assert(elementType is { });
-                        return VisitArrayOrSpanCollectionExpression(node, collectionTypeKind, node.Type, TypeWithAnnotations.Create(elementType));
+                        return VisitArrayOrSpanCollectionExpression(node, node.Type);
                     case CollectionExpressionTypeKind.CollectionBuilder:
                         // A few special cases when a collection type is an ImmutableArray<T>
-                        if (ConversionsBase.IsSpanOrListType(_compilation, node.Type, WellKnownType.System_Collections_Immutable_ImmutableArray_T, out var arrayElementType))
+                        if ((object)node.Type.OriginalDefinition == _compilation.GetWellKnownType(WellKnownType.System_Collections_Immutable_ImmutableArray_T))
                         {
-                            // For `[]` try to use `ImmutableArray<T>.Empty` singleton if available
-                            if (node.Elements.IsEmpty &&
-                                _compilation.GetWellKnownTypeMember(WellKnownMember.System_Collections_Immutable_ImmutableArray_T__Empty) is FieldSymbol immutableArrayOfTEmpty)
-                            {
-                                var immutableArrayOfTargetCollectionTypeEmpty = immutableArrayOfTEmpty.AsMember((NamedTypeSymbol)node.Type);
-                                return _factory.Field(receiver: null, immutableArrayOfTargetCollectionTypeEmpty);
-                            }
-
-                            // Otherwise try to optimize construction using `ImmutableCollectionsMarshal.AsImmutableArray`.
-                            // Note, that we skip that path if collection expression is just `[.. readOnlySpan]` of the same element type,
-                            // in such cases it is more efficient to emit a direct call of `ImmutableArray.Create`
-                            if (_compilation.GetWellKnownTypeMember(WellKnownMember.System_Runtime_InteropServices_ImmutableCollectionsMarshal__AsImmutableArray_T) is MethodSymbol asImmutableArray &&
-                                !CanOptimizeSingleSpreadAsCollectionBuilderArgument(node, out _))
-                            {
-                                return VisitImmutableArrayCollectionExpression(node, arrayElementType, asImmutableArray);
-                            }
+                            return VisitArrayOrSpanCollectionExpression(node, node.Type);
```

Comment:
shoudl this method name be renamed?

Reply:
What name would you suggest given that it handles `[Immutable]Array`s and `[ReadOnly]Span`s?